redisuniversal/redis.go

```diff
@@ -0,0 +1,229 @@
+func (c *redisStoreCore) hashTagKey(key string) string {
+	if c.clusterMode {
+		return hashTag + key
+	}
+	return key
+}
+
+// Removing the hashtag from the key
+func (c *redisStoreCore) cleanHashTagKey(key string) string {
+	if c.clusterMode {
+		return strings.Replace(key, hashTag, "",-1)
+	}
+	return key
+}
+
+func (c *redisStoreCore) GetInternal(kind ldclient.VersionedDataKind, key string) (ldclient.VersionedData, error) {
+	jsonStr, err := c.client.HGet(c.featuresKey(kind), c.hashTagKey(key)).Result()
+	if err != nil {
+		if err == redis.Nil {
+			c.loggers.Debugf("Key: %s not found in \"%s\"", key, kind.GetNamespace())
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	item, jsonErr := utils.UnmarshalItem(kind, []byte(jsonStr))
+	if jsonErr != nil {
+		return nil, fmt.Errorf("failed to unmarshal %s key %s: %s", kind, key, jsonErr)
+	}
+	return item, nil
+}
+
+func (c *redisStoreCore) GetAllInternal(kind ldclient.VersionedDataKind) (map[string]ldclient.VersionedData, error) {
+	values, err := c.client.HGetAll(c.featuresKey(kind)).Result()
+	if err != nil && err != redis.Nil {
+		return nil, err
+	}
+
+	results := make(map[string]ldclient.VersionedData)
+	for k, v := range values {
+		item, jsonErr := utils.UnmarshalItem(kind, []byte(v))
+		if jsonErr != nil {
+			return nil, fmt.Errorf("failed to unmarshal %s: %s", kind, err)
+		}
+
+		results[c.cleanHashTagKey(k)] = item
+	}
+	return results, nil
+}
+
+func (c *redisStoreCore) UpsertInternal(kind ldclient.VersionedDataKind, newItem ldclient.VersionedData) (ldclient.VersionedData, error) {
+	baseKey := c.featuresKey(kind)
+	key := newItem.GetKey()
+	var item ldclient.VersionedData
+	availableRetries := 10
+	var retryErr error
+	for availableRetries > 0 {
+		// Avoid failing in an infinite loop, we try to "upsert" te value a finite number of times,
+		//	if it can't be done after the specified retry limit, we return error
+		availableRetries--
+		err := c.client.Watch(func(tx *redis.Tx) error {
+			oldItem, err := c.GetInternal(kind, key)
+			if err != nil {
+				return err
+			}
+
+			if oldItem != nil && oldItem.GetVersion() >= newItem.GetVersion() {
+				updateOrDelete := "update"
+				if newItem.IsDeleted() {
+					updateOrDelete = "delete"
+				}
+				c.loggers.Debugf(`Attempted to %s key: %s version: %d in "%s" with a version that is the same or older: %d`,
+					updateOrDelete, key, oldItem.GetVersion(), kind.GetNamespace(), newItem.GetVersion())
+				item = oldItem
+				return nil
+			}
+
+			data, jsonErr := json.Marshal(newItem)
+			if jsonErr != nil {
+				return fmt.Errorf("failed to marshal %s key %s: %s", kind, key, jsonErr)
+			}
+
+			result, err := tx.TxPipelined(func(pipe redis.Pipeliner) error {
+				err = pipe.HSet(baseKey, c.hashTagKey(key), data).Err()
+				if err == nil {
+					result, err := pipe.Exec()
+					// if exec returned nothing, it means the watch was triggered and we should retry
+					if err == nil && len(result) > 0 {
```

Comment:
I believe this was meant to be `len(result) == 0`. As the comment says, a zero-length result means that the watch was triggered, so a non-zero length means that it was not and so the update succeeded.

Running the standard SDK database test suite, as all of the existing database integrations do, would have caught this.